pkg/liqonet/utils_test.go

```diff
@@ -0,0 +1,124 @@
+
+	Describe("Getting ClusterID info from ConfigMap", func() {
+
+		var (
+			ns             string
+			clusterIDKey   string
+			clusterIDValue string
+			configMapData  map[string]string
+			configMap      corev1.ConfigMap
+			backoff        wait.Backoff
+		)
+
+		BeforeEach(func() {
+			ns = "default"
+			clusterIDKey = "cluster-id"
+			clusterIDValue = "fake-cluster-id"
+			configMapData = make(map[string]string, 1)
+			configMapData[clusterIDKey] = clusterIDValue
+			configMap = corev1.ConfigMap{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ConfigMap",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      clusterIDKey,
+					Namespace: ns,
+				},
+				Data: configMapData,
+			}
+
+			// 3 attempts with 15 seconds sleep between one another
+			// for a total of 30 seconds
+			backoff = wait.Backoff{
+				Steps:    3,
+				Duration: 15 * time.Second,
+				Factor:   1.0,
+				Jitter:   0,
+			}
+		})
+
+		Context("When it has not been created yet", func() {
+			It("should retry more times and eventually fail", func() {
+				start := time.Now()
+				_, err := liqonet.GetClusterID(clientset, clusterIDKey, ns, backoff)
+				end := time.Now()
+				Expect(err).To(HaveOccurred())
+
+				timeout := backoff.Duration * time.Duration(backoff.Factor-1)
+				Expect(end).Should(BeTemporally(">=", start.Add(timeout)))
+			})
+		})
+		Context("When it has been already created", func() {
+			It("should return immediately the clusterID value", func() {
+				_, err := clientset.CoreV1().ConfigMaps(ns).Create(
+					context.TODO(),
+					&configMap,
+					metav1.CreateOptions{},
+				)
+				Expect(err).NotTo(HaveOccurred())
+				start := time.Now()
+				clusterID, err := liqonet.GetClusterID(clientset, clusterIDKey, "default", backoff)
+				end := time.Now()
+				Expect(err).NotTo(HaveOccurred())
+				Expect(clusterID).To(Equal(clusterIDValue))
+
+				Expect(end).Should(BeTemporally("<", start.Add(backoff.Duration)))
+
+				// clean
+				err = clientset.CoreV1().ConfigMaps(ns).Delete(
+					context.TODO(),
+					clusterIDKey,
+					metav1.DeleteOptions{},
+				)
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
+		Context("When it has been created during backoff", func() {
+			It("should detect the configmap and return the ClusterID value", func() {
+
+				go func() {
+					time.Sleep(25 * time.Second)
+					_, err := clientset.CoreV1().ConfigMaps(ns).Create(
+						context.TODO(),
+						&configMap,
+						metav1.CreateOptions{},
+					)
+					Expect(err).NotTo(HaveOccurred())
+				}()
+				start := time.Now()
+				clusterID, err := liqonet.GetClusterID(clientset, clusterIDKey, "default", backoff)
+				end := time.Now()
+				Expect(err).NotTo(HaveOccurred())
+
+				minimum_sleep := 2 * backoff.Duration
+				Expect(end).Should(BeTemporally(">=", start.Add(minimum_sleep)))
+				Expect(clusterID).To(Equal(clusterIDValue))
+
+				// clean
+				err = clientset.CoreV1().ConfigMaps(ns).Delete(
```

Comment:
Since this code is duplicated, I would embed the 2nd and the 3rd contexts in a Describe function with an AfterEach statement that cleans up the created configmap

EDIT: If you'd use the fake clientset instead of the controller-runtime test environment, the explicit delete is not needed at all, because the client is instantiated for each test case.